sdr_uhd.c

```diff
@@ -0,0 +1,208 @@
+// Part of dump1090, a Mode S message decoder for RTLSDR devices.
+//
+// sdr_bladerf.h: UHD support (header)
+//
+// Copyright (c) 2017 FlightAware LLC
+//
+// This file is free software: you may copy, redistribute and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 2 of the License, or (at your
+// option) any later version.
+//
+// This file is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "dump1090.h"
+#include "sdr_uhd.h"
+
+#include <inttypes.h>
+
+// complex int16
+#define SAMPLE_SIZE 4
+
+void free_usrp();
+void free_device_args();
+void free_rx_metadata();
+void free_rx_streamer();
+
+
+static struct {
+    uhd_usrp_handle usrp;
+    uhd_rx_streamer_handle rx_streamer;
+    uhd_rx_metadata_handle md;
+    char *device_args;
+    size_t free_func_ptr;
+    size_t channel;
+    char *sample_data;
+
+    iq_convert_fn converter;
+    struct converter_state *converter_state;
+} uhd;
+
+
+void (*free_func[]) (void) = {
+    NULL,
+    free_device_args,
+    free_usrp,
+    free_rx_streamer,
+    free_rx_metadata,
+};
+
```

Comment:
Not sure I am a fan of this -- now you have to make sure to keep the initialization code (and the magic points at which free_func_ptr is incremented) with this array.

Can you do the traditional "goto error_xyz" with fallthrough for error handling instead perhaps?

Reply:
Done.